### PR TITLE
Draft order changes

### PR DESCRIPTION
**Pull request by shashisk97**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
PRL-1643

### Change description ###
Solicitor draft order

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes
[ ] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -540,6 +543,7 @@ def sonarExclusions = [
   '**uk/gov/hmcts/reform/prl/utils/ElementUtils.java',
   '**uk/gov/hmcts/reform/prl/services/validators/EventsChecker.java',
   '**uk/gov/hmcts/reform/prl/services/citizen/*',
+  '**uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java'
```
Comment (AlokDatta) [inline build.gradle:546]: Should we remove this now?

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        CaseData caseData = objectMapper.convertValue(
+            callbackRequest.getCaseDetails().getData(),
+            CaseData.class
+        );
+        if (PrlAppsConstants.FL401_CASE_TYPE.equalsIgnoreCase(caseData.getCaseTypeOfApplication())) {
+            callbackRequest.getCaseDetails().getData().putAll(draftAnOrderService.populateCustomFields(caseData).toMap(
+                CcdObjectMapper.getObjectMapper()));
+        }
+
+        log.info("*** caseDataUpdated {} ***", callbackRequest.getCaseDetails().getData());
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:74]: hide this logger please.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+
+    @PostMapping(path = "/solicitor-prepopulate-draft-order-text", consumes = APPLICATION_JSON, produces = APPLICATION_JSON)
+    @Operation(description = "Callback to Generate text for draft an order")
+    public AboutToStartOrSubmitCallbackResponse prepopulateSolicitorDraftAnOrder(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseData caseData = CaseUtils.getCaseData(callbackRequest.getCaseDetails(), objectMapper);
+        log.info("*** inside prepopulateSolicitorDraftAnOrder() {} ***", caseData);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:85]: this one too.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        callbackRequest.getCaseDetails().getData().put(
+            "previewDraftAnOrder",
+            draftAnOrderService.getTheOrderDraftString(caseData)
+        );
+        log.info("*** before returning {} ***", callbackRequest.getCaseDetails().getData());
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:90]: another instance

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    @PostMapping(path = "/about-to-submit", consumes = APPLICATION_JSON, produces = APPLICATION_JSON)
+    @Operation(description = "Callback to generate draft order collection")
+    public AboutToStartOrSubmitCallbackResponse prepareDraftOrderCollection(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseData caseData = CaseUtils.getCaseData(callbackRequest.getCaseDetails(), objectMapper);
+        Map<String, Object> caseDataUpdated = callbackRequest.getCaseDetails().getData();
+        caseDataUpdated.putAll(draftAnOrderService.generateDraftOrderCollection(caseData));
+        log.info("*** before returning {} ***", caseDataUpdated);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:102]: remove the logger.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        return AboutToStartOrSubmitCallbackResponse.builder().data(caseDataUpdated).build();
+    }
+
+    @PostMapping(path = "/generate-draft-an-order", consumes = APPLICATION_JSON, produces = APPLICATION_JSON)
+    @Operation(description = "Callback to Generate document for draft an order")
+    public AboutToStartOrSubmitCallbackResponse draftAnOrderMidEventCallback(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) throws Exception {
+        log.info("request prop {}", callbackRequest.getCaseDetails().getData());
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:111]: logger.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        CaseData caseData = CaseUtils.getCaseData(callbackRequest.getCaseDetails(), objectMapper);
+        log.info("caseData {}", caseData);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:113]: logger.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        Map<String, Object> caseDataUpdated = new HashMap<>();
+        caseDataUpdated.put(
+            "solicitorOrJudgeDraftOrderDoc",
+            draftAnOrderService.generateSolicitorDraftOrder(authorisation, caseData)
+        );
+        log.info("*** caseDataUpdated {} ***", caseDataUpdated);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/DraftAnOrderController.java:119]: logger

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/EditAndApproveDraftOrderController.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            "********caseData.getDoYouWantCourtAdminToAddAnything***** {}",
+            caseData.getDoYouWantCourtAdminToAddAnything()
+        );
+        if (callbackRequest.getEventId().equalsIgnoreCase("adminEditAndApproveAnOrder")) {
+            caseDataUpdated.putAll(draftAnOrderService.removeDraftOrderAndAddToFinalOrder(authorisation, caseData));
+        } else {
+            caseDataUpdated.putAll(draftAnOrderService.updateDraftOrderCollection(caseData));
+        }
+
+        log.info("*** before returning {} ***", caseDataUpdated);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/EditAndApproveDraftOrderController.java:86]: logger

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/EditAndApproveDraftOrderController.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    public AboutToStartOrSubmitCallbackResponse generateAdminJudgeDraftOrder(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) throws Exception {
+        CaseData caseData = objectMapper.convertValue(
+            callbackRequest.getCaseDetails().getData(),
+            CaseData.class
+        );
+        Map<String, Object> caseDataUpdated = new HashMap<>();
+        caseDataUpdated.put(
+            "solicitorOrJudgeDraftOrderDoc",
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/EditAndApproveDraftOrderController.java:117]: constant?

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+                    ? "out" : ""
+            );
+
+            StringSubstitutor substitutor = new StringSubstitutor(nonMolestationPlaceHoldersMap);
+            return substitutor.replace(nonMolestationOrderString);
+        }
+        return null;
+    }
+
+    private String getRespondentMustNotOrders(FL404 fl404CustomFields) {
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:301]: Must Not Orders? what does it mean?

Comment (AlokDatta) [reply src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:301]: Got it, lets put the order name here

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 12]
```diff
@@ -0,0 +1,736 @@
...
+                break;
+
+            default:
+                break;
+
+        }
+        return temp;
+    }
+
+    private String getNonMolestationString(String nonMolestationOrderString, CaseData caseData) {
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:121]: Lets put the order name here?

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 13]
```diff
@@ -0,0 +1,736 @@
...
+
+    private DraftOrder getUpdatedDraftOrder(DraftOrder draftOrder, CaseData caseData) {
+
+        return DraftOrder.builder().orderType(draftOrder.getOrderType())
+            .orderTypeId(draftOrder.getOrderTypeId())
+            .orderDocument(caseData.getSolicitorOrJudgeDraftOrderDoc())
+            .otherDetails(OtherDraftOrderDetails.builder()
+                              .createdBy(draftOrder.getOtherDetails().getCreatedBy())
+                              .dateCreated(dateTime.now())
+                              .status("JUDGE_COMMENTS_ENTERED").build())
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:484]: Move this to an ENUM or constant?

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+        if (draftOrder.getJudgeNotes() != null) {
+            caseDataMap.put("instructionsFromJudge", draftOrder.getJudgeNotes());
+        }
+        log.info("inside populateSelectedOrder {}", caseDataMap);
+        return caseDataMap;
+    }
+
+
+    public Map<String, Object> populateSelectedOrderText(CaseData caseData) {
+        log.info("inside populateSelectedOrderText caseData {}", caseData);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:529]: logger

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+        Map<String, Object> caseDataMap = new HashMap<>();
+        caseDataMap.put("previewDraftAnOrder", getDraftOrderText(caseData));
+        log.info("inside populateSelectedOrderText {}", caseDataMap);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:532]: logger

File: src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java [hunk 16]
```diff
@@ -0,0 +1,736 @@
...
+                uk.gov.hmcts.reform.prl.models.dto.ccd.CaseDetails
+                    .builder().caseData(caseData).build(),
+                solicitorFinalOrder
+            );
+        } catch (Exception e) {
+            log.error("Error while generating the final document");
+        }
+        log.info("*************courtAdminNotes {}", caseData.getPreviewDraftAnOrder());
+        log.info("***************draftorder {}", draftOrder);
+        log.info("*************courtAdminNotes {}", caseData.getCourtAdminNotes());
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/DraftAnOrderService.java:619]: remove this loggers.

**Review by AlokDatta — COMMENTED** [review]
Can we tag this repo against Release 3 branch please.

**Review by AlokDatta — DISMISSED** [review]

**Comment by BharadwajSairam** [comment]
Closing this PR since editable text area approach is halted for solicitor draft order

**Pull request closed without merging**